lib/src/tree_builder.rs

```diff
@@ -55,19 +55,22 @@ impl TreeBuilder {
     }
 
     pub fn set(&mut self, path: RepoPath, value: TreeValue) {
+        assert!(!path.is_root());
```

Comment:
I'm not sure about this constraint because it seems like a reasonable thing for e.g. `jj restore --from X .` to do (though it doesn't currently have that optimization for setting tree values). I don't mind having the constraint here for now; I just think we may want to relax it in the future.

Reply:
I consider `TreeBuilder` doesn't practically support directory path to set/remove(). We can set tree id to override, but no subtree items should be set/removed in that case. So I think `.set(root, tree_id)` is no different than simply doing `return tree_id`.

--
I found one more `path.split()` which can be unwrapped, so I'll update the patch.

Reply:
> So I think `.set(root, tree_id)` is no different than simply doing `return tree_id`.

What I meant is that it might make sense in generic code. Let's say `jj restore` had an optimization to call `set(path, tree_id)` when the matcher returned `AllRecursively`. Such an optimization would make sense for `jj restore --from X some/dir`, I think. If we had that optimization, then it would also apply to the root directory. Also, it would mean that we could treat full restore (no paths on the CLI) the same as partial restore with `matcher = EverythingMatcher`.

Reply:
Yeah, that's a valid use case. I just think it's scary for TreeBuilder to support `set(path, tree_id)` with the current form. We might need to ensure `.set("foo/bar",)` fails if "foo" is overridden by new tree, `.remove()` might have to drop subtree overrides, etc.

Reply:
Yes, sounds good. That can be done later. Feel free to merge this as is.